Views/BuddyView: Introduce BuddyView to show Recent PMs and Group PMs.

amanagr opened the pull request.

amanagr committed:
TopButton: Use 'any' wrap for overflowing texts.

Don't space wrap the last word if it doesn't fit in the column.
Instead fill part of word in one row and rest in the next row.
This is done to keep computation of showing unread count at
the right position simple. Otherwise, we will have to take
wrapping of words into account which is currently not included
in the code.

amanagr committed:
helper/index_messages: Store all the PM recipients in index.

Store PM recipients in index as dict of frozenset(recipients)
and Timestamp to be used in BuddyView.

amanagr committed:
test_classify_unread_counts: Add tests for classify_unread_counts.

Add tests and fixtures for testing classify_unread_counts.

amanagr committed:
BuddyView: Implement v1 of buddy list.

This includes following features:
 * `b` keypress toggling of BuddyView
 * Shows Recent PMs and Group PMs
 * Shows corresponding Unread Counts.
 * Shows Online/Idle/Offline status of users/Group PMs.

core/keys: Use 'b' to toggle BuddyView.
Fixes #220

amanagr committed:
buttons/TopButton: Don't shrink text in case of group pms.

Allow caption to overflow to next line if `shrink=False`.

amanagr committed:
GroupPMButton: Introduce Button to show GroupPMs.

This is based on TopButton except that it has multiple recipients.
Add tests to check correctly rendered caption.

amanagr commented:
![buddy_view](https://user-images.githubusercontent.com/25124304/53560921-7d3df100-3b73-11e9-87bb-0f772224b01b.png)
This is a preview of how buddy list looks now. This is mostly functional and is ready for a spin if anyone wants to try. I would love any feedback on this. 

I will add tests for this soon and then reorder some of the commits.

amanagr commented:
I don't know why the tests are failing for python 3.4 only??

amanagr commented:
@neiljp I have added minimal tests for the functions and this could possibly serve as a v1 for BuddyList. Since this PR is already getting too big, I will add some feature improvements I have in mind along with more robust tests in future PRs.  Thanks!

neiljp commented on zulipterminal/helper.py:
```diff
@@ -341,6 +340,14 @@ def classify_unread_counts(model: Any) -> UnreadCounts:
+        user_ids = frozenset(map(int, user_ids))
+        unread_counts['unread_pms'][user_ids] = count
+        unread_counts['all_msg'] += count
+        unread_counts['all_pms'] += count
```
The rest of this code builds infrastructure for testing classify_unread_counts - which is great :)

However, I would suggest putting the above additions and the changes to the tests required for it from this commit in a separate commit.

Both of those commits would then be fine in another PR?

neiljp commented on tests/conftest.py:
```diff
@@ -350,11 +350,112 @@ def initial_data(logged_on_user):
         'unread_msgs': {
-            'pms': [],
-            'count': 0,
+            'pms': [{
```
I like the addition of this extra data to the tests - but this is a lot. Is this real data you've captured? Can we trim it to base data and then parametrize to look at some cases? This looks like a lot, and while having real data is good, it makes it more difficult to understand?

neiljp reviewed (commented):
I don't have much time to do a complete review as this is quite large and mixed, but I've left some thoughts on the first and third, in order to give feedback sooner rather than later.

While I'm unsure right now if the current UI is the approach to go, I do like that this builds UI elements that we could use wherever we want; I hope the inheritance helps?

amanagr commented:
Thanks @neiljp for the review I will get to them by the end of this week as I am busy with exams.

amanagr commented on tests/ui/test_ui_tools.py:
```diff
@@ -1244,6 +1244,7 @@ def test_text_content(self, mocker,
                 count_str)
         assert len(text[0]) == len(expected_text) == (width - 1)
         assert text[0] == expected_text
+        assert top_button._w._original_widget._wrap_mode == 'any'
```
Primarily, to keep computation of showing unread count at the right position simple. Otherwise, we will have to take wrapping of words into account which is currently not included in the code.

neiljp replied:
I can see that this could confuse the placement of the unread counts. That said, the styling of the words looks pretty bad if the text wraps mid-word.

I would put this 'why' into the commit title/summary in any case.

amanagr replied:
Of course! done :+1:

amanagr committed:
classify_unread_counts: Store unread of count of GroupPMs.

Get unread_counts for group pms stored as 'huddles' and store
them in unread_msg_counts.
Add tests for classify_unread_counts along with appropriate
fixtures.

amanagr commented:
I have split the commit as you suggested. As for the fixtures, I can curb them down but I feel they can be useful later as they provide the base for all the group_pm and unread_count related fixtures which were previously missing.

amanagr commented:
@neiljp is this ready to merge?

amanagr commented:
@punchagan @neiljp I have rebased this PR against master. Let me know your views on this.

amanagr committed:
unread_count: Set correct count for the group_pms.

amanagr commented:
Closing since this is unlikely to be useful point to start from.
